pkg/util/docker/event_stream.go

```diff
@@ -0,0 +1,115 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+// +build docker
+
+package docker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+)
+
+//// Can't be unit tested, covered by the listeners/docker
+//// and dogstatsd/origin_detection integration tests.
+
+// eventSuscriber holds the state for a subscriber
+type eventSuscriber struct {
+	name       string
+	eventChan  chan *ContainerEvent
+	errorChan  chan error
+	cancelChan chan struct{}
+}
+
+// SubscribeToContainerEvents allows a package to suscribe to events from the event stream.
+// A unique subscriber name should be provided.
+func (d *DockerUtil) SubscribeToContainerEvents(name string) (<-chan *ContainerEvent, <-chan error, error) {
+	d.RLock()
+	if _, found := d.eventSuscribers[name]; found {
+		d.RUnlock()
+		return nil, nil, errors.New("already suscribed")
+	}
+	d.RUnlock()
+
+	sub := &eventSuscriber{
+		name:       name,
+		eventChan:  make(chan *ContainerEvent, 5),
+		errorChan:  make(chan error, 5),
+		cancelChan: make(chan struct{}),
+	}
+	d.Lock()
+	d.eventSuscribers[name] = sub
+	d.Unlock()
+
+	go d.streamEvents(sub.eventChan, sub.cancelChan)
+	return sub.eventChan, sub.errorChan, nil
+}
+
+// UnsuscribeFromContainerEvents allows a package to unsuscribe.
+// The call is blocking until the request is processed.
+func (d *DockerUtil) UnsuscribeFromContainerEvents(name string) error {
+	d.Lock()
+	sub, found := d.eventSuscribers[name]
+	if !found {
+		d.Unlock()
+		return errors.New("not suscribed")
+	}
+	delete(d.eventSuscribers, name)
+	d.Unlock()
+
+	// Block until the goroutine exists, then close all chans
+	sub.cancelChan <- struct{}{}
+	close(sub.cancelChan)
+	close(sub.errorChan)
+	close(sub.eventChan)
+
+	return nil
+}
+
+func (d *DockerUtil) streamEvents(dataChan chan<- *ContainerEvent, cancelChan <-chan struct{}) {
+	filters := filters.NewArgs()
+	filters.Add("type", "container")
+	filters.Add("event", "start")
+	filters.Add("event", "die")
+
+	// Outer loop handles re-connecting in case the docker daemon closes the connection
+	for {
+		eventOptions := types.EventsOptions{
+			Since:   fmt.Sprintf("%d", time.Now().Unix()),
```

Comment:
I may prefer a `strconv.FormatInt` like instead of formatting, what do you think ?